### DISCUSSION TITLE
Extending or using kupo as a library?

### DISCUSSION THREAD

**Discussion by cortsf** [CardanoSolutions/kupo#74] [Q&A] [answered]
Hello everybody. Do you think it would be hard to write a project-specific custom indexer based of kupo? Basically we need:

1. Paginated results (or a simple limit)
2. Indexed metadata
3. Precomputed value storage (this needs to be kept up to date by the indexer)

We are considering many solutions, since we are a team of haskeller we would like to choose kupo if we have to fork an existing indexer.

Thanks!

**Comment by KtorZ** [comment, marked as answer]
Hey, thanks for reaching out, and glad to see you're considering Kupo ☺️

Regarding your questions:

1. Pagination, yes. This is something I've discussed on the side with various users already and that is coming. Thanks for opening a ticket about this, I'll dump some thoughts and inform about the direction this is headed. If it also solves your use-case, that'd be great.

2. Kupo does provide access to transaction metadata (although only on the nightly builds, for the upcoming v2.1.0). I am not sure if you haven't seen this or, if the current approach doesn't solve your use-case? If it doesn't, I'd be happy to understand what it is in more detail.

3. This sounds like something [Scrolls]( https://github.com/txpipe/Scrolls) would be good at. Kupo, possibly but hard to answer without more details on what does that mean exactly? I assume 'Value' here does not refer to output values, but refer to the more generic sense of it.

Finally, regarding forking and extending in your way, that should be possible, especially if you're already familiar/proficient with Haskell. Kupo is mostly written using boring Haskell. Yet, the internal engine has a few parts which may be harder to grasp -- I have arguably underdocumented the project's architecture yet, I am happy to take the time to do so. I've been planning on it for a while for had no strong compelling reason to do it so far.

**Comment by cortsf** [comment]
Thanks for the answer @KtorZ. In the end we are choosing oura. Both are excelent projects, and I personally love the simplicity of Kupo (and haskell :smile: ). But our requirements are a bit weird and we still need oura for notification so we are writting a custom indexer on top of it. 

Thank you, and keep the good work! After having done a lot on research on this topic, I can say the quality of the documentation, and the simplicity of getting started with your tools is by far the best out there.
